content/Refactoring - Sebastian Herrera.md

```diff
@@ -0,0 +1,144 @@
+
+After reaching to a really much better solution than the original, I really liked what we had done, and I saw a lot of potential in this technique. A code that already worked, now worked the same, but it was much easier to understand and modify whenever it was needed.
+
+Last year, I felt that I definitely could keep improving my skills, so after searching for books recommendations, I reached one very well known: "Refactoring" by Martin Fowler. Now I can say that I understand why it is so recommended. This topic I became interested long ago, now made much more sense to me.
+
+For you interested in the origins of things, refactoring wasn’t invented by Martin Fowler as some people might think. He says that he learned this technique from Kent Beck and a group of developers of a community of Smalltalk language back in the 90s, and that by that time there were just a few articles and thesis about it, but there was no book.
+
+So, enough introduction, this is about what I learned from that book.
+
+## 1) WHY TO DO REFACTORING?
+Maybe the most important question to ask yourself before doing anything, is why should you do it, meaning what value does it give in return.
+
+These are a some of the main reasons why doing refactoring is recommended:
+- Refactoring improves the design of software.
+  - Creating software that has a good architecture isn't something that can be done first time. The design we believe that best fits today, might not be the one that best fits next week.
+  - Without refactoring the design of the program will decay over time.
+- Refactoring makes software easier to understand.
+  - *“Any fool can write code that a computer can understand. Good programmers write code that humans can understand.”* - Martin Fowler
+- Refactoring makes you find bugs.
+  - When the code is more simple, it is easier to understand, easier to debug, and also easier to detect non-considered cases.
+- Refactoring helps you program faster.
+  - Without a good design, you can progress quickly for a while, but soon the poor design starts to slow you down.
+
+**Conclusion: Refactoring lowers the cost of enhancements.**
+
+## 2) WHEN TO DO REFACTORING?
+Refactoring isn't a special task that will show up in a project plan, so when should you do it?
+- When you need to add a new feature to a codebase, look at the existing code and consider whether it's structured in such a way to make the new change straightforward.
+  - If it isn't, then refactor the existing code to make this new addition easy.
+  - Once that change is done, then add the new feature.
+- After you add a feature and get it working, you can a review the resulting code, that while it works, it may not be as clear as it could be.
+  - If it isn't , then refactor it into a better shape so that when you (or someone else) return to this code in a few weeks time, you won't have to spend time puzzling out how this code works.
```

Comment:
```suggestion
  - If it isn't, then refactor it into better shape so that when you (or someone else) return to this code in a few weeks, you won't have to spend time puzzling out how this code works.
```